File: examples/python/run_async_model_wait.py
#
# Copyright (C) 2018- DEEPX Ltd.
# All rights reserved.
#
# This software is the property of DEEPX and is provided exclusively to customers
# who are supplied with DEEPX NPU (Neural Processing Unit).
# Unauthorized sharing or usage is strictly prohibited by law.
#

import numpy as np
import argparse
import os
import time
from dx_engine import InferenceEngine
from logger import Logger, LogLevel

import queue
import threading

q = queue.Queue()

def inference_thread_func(ie, loop_count):
    logger = Logger()
    count = 0

    while(True):

        # pop item from queue
        jobId = q.get() # NOSONAR

        # waiting for the inference to complete by jobId
        # ownership of the outputs is transferred to the user
        outputs = ie.wait(jobId) # NOSONAR:python:S1481

        # post processing
        # postProcessing(outputs);

        # something to do


        logger.debug(f"Inference outputs corresponding to jobId={jobId}, index={count}")

        count += 1
        if ( count >= loop_count ):
            break

    return 0


def parse_args():
    parser = argparse.ArgumentParser(description="Run asynchronous model inference with wait")
    parser.add_argument("--model", "-m", type=str, required=True, help="Path to model file (.dxnn)")
    parser.add_argument("--loops", "-l", type=int, default=1, help="Number of inference loops (default: 1)")
    parser.add_argument("--verbose", "-v", action="store_true", default=False, help="Enable debug logging")
    args = parser.parse_args()

    if not os.path.exists(args.model):
        parser.error(f"Model path '{args.model}' does not exist.")

    if args.verbose:
        logger = Logger()
        logger.set_level(LogLevel.DEBUG)

    return args


if __name__ == "__main__":
    args = parse_args()
    logger = Logger()

    logger.info(f"Start run_async_model_wait test for model: {args.model}")

    try:
        # create inference engine instance with model
        with InferenceEngine(args.model) as ie:

            # do not register call back function
            # ie.register_callback(on_inference_callback_func)

            t1 = threading.Thread(target=inference_thread_func, args=(ie, args.loops))

            t1.start()

            # NOTE: np.zeros() uses COW zero pages — all virtual pages share one
            # physical page. PCIe DMA driver's get_user_pages() then sees duplicate
            # physical pages in the SG list and fails with EFAULT.
            # np.empty() + explicit fill forces unique physical page allocation.
            _buf = np.empty(ie.get_input_size(), dtype=np.uint8)
            _buf.fill(0)
            input = [_buf]
            start = time.perf_counter()

            # inference loop
            for i in range(args.loops):


                # inference asynchronously, use all npu cores
                # if device-load >= max-load-value, this function will block
                jobId = ie.run_async(input, user_arg=0)

                q.put(jobId)

                logger.debug(f"Inference start (async) {i}")

            t1.join()

            end = time.perf_counter()
            total_time_ms = (end -start) * 1000
            avg_latency = total_time_ms / args.loops
            fps = 1000.0/ avg_latency if avg_latency > 0 else 0.0

            logger.info("-----------------------------------")
            logger.info(f"Total Time: {total_time_ms:.3f} ms")
            logger.info(f"Average Latency: {avg_latency:.3f} ms")
            logger.info(f"FPS: {fps:.2f} frame/sec")
            logger.info("Success")
            logger.info("-----------------------------------")

    except Exception as e:
        logger.error(f"Exception: {str(e)}")
        exit(-1)

    exit(0)

File: examples/cpp/run_async_model_wait/run_async_model_wait.cpp
/*
 * Copyright (C) 2018- DEEPX Ltd.
 * All rights reserved.
 *
 * This software is the property of DEEPX and is provided exclusively to customers
 * who are supplied with DEEPX NPU (Neural Processing Unit).
 * Unauthorized sharing or usage is strictly prohibited by law.
 */

#include "dxrt/dxrt_api.h"
#include "dxrt/extern/cxxopts.hpp"
#include "../include/logger.h"
#include "../include/concurrent_queue.h"

#include <string>
#include <iostream>


// concurrent queue is a thread-safe queue data structure
// designed to be used in a multi-threaded environment
static ConcurrentQueue<int> gJobIdQueue(32);

// user thread to wait for the completion of inference
static int inferenceThreadFunc(const dxrt::InferenceEngine& ie, int loopCount)
{
    static const auto& log = dxrt::Logger::GetInstance();
    int count = 0;

    while(true)
    {
        // pop item from queue
        int jobId = gJobIdQueue.pop();

        try
        {
            // waiting for the inference to complete by jobId
            // ownership of the outputs is transferred to the user
            auto outputs = ie.Wait(jobId);

            // now there is no post processing
            (void)outputs;

            // something to do
        }
        catch (const dxrt::Exception& e)
        {
            log.Error(std::string(e.what()) + " error-code=" + std::to_string(static_cast<int>(e.code())));
            return -1;
        }

        log.Debug("Inference outputs corresponding to jobId(" + std::to_string(jobId) + ")");

        count++;
        if ( count >= loopCount ) break;

    } // while

    return 0;
}

int main(int argc, char* argv[])
{
    std::string model_path;
    int loop_count;
    bool verbose;

    auto &log = dxrt::Logger::GetInstance();

    cxxopts::Options options("run_async_model_wait", "Run asynchronous model inference with wait");
    options.add_options()
        ("m,model", "Path to model file (.dxnn)", cxxopts::value<std::string>(model_path))
        ("l,loops", "Number of inference loops", cxxopts::value<int>(loop_count)->default_value("1"))
        ("v,verbose", "Enable verbose/debug logging", cxxopts::value<bool>(verbose)->default_value("false"))
        ("h,help", "Print usage");

    try
    {
        auto result = options.parse(argc, argv);

        if (result.count("help") || !result.count("model"))
        {
            std::cout << options.help() << std::endl;
            return result.count("help") ? 0 : -1;
        }

        if (verbose) {
            log.SetLevel(dxrt::Logger::Level::LOGLEVEL_DEBUG);
        }
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("Error parsing arguments: ") + e.what());
        std::cout << options.help() << std::endl;
        return -1;
    }

    log.Info("Start async_model_wait test for model: " + model_path);

    try
    {
        // create inference engine instance with model
        dxrt::InferenceEngine ie(model_path);

        // do not register call back function
        // inferenceEngine.RegisterCallback(onInferenceCallbackFunc);

        // create thread
        auto t1 = std::thread(inferenceThreadFunc, std::ref(ie), loop_count);

        // create temporary input buffer for example
        std::vector<uint8_t> inputPtr(ie.GetInputSize(), 0);

        auto start = std::chrono::high_resolution_clock::now();

        // inference loop
        for(int i = 0; i < loop_count; ++i)
        {
            // no need user argument
            // UserData *userData = getUserDataInstanceFromDataPool();

            // inference asynchronously, use all npu cores
            // if device-load >= max-load-value, this function will block
            auto jobId = ie.RunAsync(inputPtr.data());

            // push jobId in global queue variable
            gJobIdQueue.push(jobId);

            log.Debug("Inference request submitted with jobId(" + std::to_string(jobId) + ")");
        } // for i

        t1.join();

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration = end - start;

        double total_time = duration.count();
        double avg_latency = total_time / static_cast<double>(loop_count);
        double fps = 1000.0 / avg_latency;

        log.Info("-----------------------------------");
        log.Info("Total Time: " + std::to_string(total_time) + " ms");
        log.Info("Average Latency: " + std::to_string(avg_latency) + " ms");
        log.Info("FPS: " + std::to_string(fps) + " frame/sec");
        log.Info("Success");
        log.Info("-----------------------------------");
    }
    catch (const dxrt::Exception& e)
    {
        log.Error(std::string("dxrt::Exception: ") + e.what());
        return -1;
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("std::exception: ") + e.what());
        return -1;
    }
    catch(...)
    {
        log.Error("Exception");
        return -1;
    }

    return 0;
}

